pages/blog/websocket-part3.md

```diff
@@ -0,0 +1,605 @@
+
+### AsyncAPI 2.0 Limitations
+
+As I mention in this article and the previous ones on WebSocket, you write AsyncAPI document for your application from a client perspective. When you start designing your API, you ask yourself questions from the user's perspective.
+
+Don't get me wrong; I'm not saying that taking user perspective is wrong. It makes sense but also confusing for many. I recommend joining [this](https://github.com/asyncapi/spec/issues/520) thread for more details.
+
+Confusion between **subscribe** and **publish** is not the only problem. Once you understand that your application that publishes events must describe it as subscribe operation, the rest is trivial.
+
+The real problem is with code generation. You describe the application from a user perspective, so client code generation is easy. What about generating code for your application. Have a look again at the **/travel/status** channel from ShrekApp:
+
+```yml
+  /travel/status:
+    subscribe:
+      summary: Client can receive travel info status.
+      operationId: subTravelInfo
+      message:
+        $ref: '#/components/messages/travelInfo'
+```
+
+I had to try hard to make sure the **summary** and **operationID** are as a neutral as possible. My first version looked like this:
+
+```yml
+  /travel/status:
+    subscribe:
+      summary: You can listen to travel info status.
+      operationId: onTravelInfo #client code perspective, generated client reacts "onTravelInfo" incomming message 
+      message:
+        $ref: '#/components/messages/travelInfo'
+```
+
+Then in **src/api/services/travel-status.js** file, I would have **onTravelInfo** instead of **subTravelInfo** with the following jsdoc:
+
+```js
+/**
+ * You can listen to travel info status.
+ * @param {object} ws WebSocket connection.
+ */
+service.onTravelInfo = async (ws) => {
+...
+```
+
+This is a function responsible for sending messages to the Client once it connects with a given channel. Function name like **onTravelInfo** is misleading, not to mention the generated code description.
+
+I chose neutral descriptions. They are acceptable, I think, but not when it comes to user-facing documentation. Every technical writer will tell you that the best docs are the ones that are directed to a reader. You don't write "what user can do with the API" but "what you can do with the API".
+
+Therefore, all descriptions and even things like **operationId** should have two versions to satisfy both docs and code depending on the perspective. Otherwise, you need to make a sacrifice. You either make the developer that maintains the code happy or the technical writer that maintains docs happy.
+
+In my opinion, this limitation should not block you from choosing the direction of API-first. The benefits are too big to resign that easily. Just join us and let us find solutions together. We want to help solve all those issues, but we just need some help from you too.
```

Comment:
I think you should create an issue in the spec. Let's make sure we discuss it. And if you do it before publishing this article, you can include a link to the issue here.

Reply:
for me this is part of this https://github.com/asyncapi/spec/issues/520 discussions, it is all about the "perspective", the thing is that there is just one, and therefore issue with descriptions and operationId, maybe I'll just add a comment there as publish/subscribe fix should take into consideration all aspects of the "perspective" appraoch

Reply:
Sorry, I meant the paragraph before 🤦 The one about `operationId`.

Reply:
@fmvilas added